### PR TITLE
Allison Northrop - RideShare - Queues 

### PR DESCRIPTION
**Pull request by Allison-Northrop**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |  A design decision I made included determining what methods to make self methods and what methods to make instances methods. For example, for anything that involved outputting the csv file, I made it a self method, because it was related to the entire class and not just an instance of a class. I also was able to use that data in other classes. For example, in the Driver class, I made the find_driver(driver_id) a self method so that I was able to use that method as a starting point in the Trip class's driver_object method       |
| Describe a concept that you gained more clarity on as you worked on this assignment. | I gained way more clarity on self methods. I finally realized that self methods don't even technically have to be in the class (as in they have nothing to do with one instance of the class). They're just methods that are related to a lot of instances of the classes as opposed to one instance of the class.  |
| Describe a _nominal test_ that you wrote for this assignment. |  An example of a nominal test I wrote for this assignment is in the driver spec I checked if it could find/return a driver account that existed for the find_drivers(driver_id) method.  |
| Describe an _edge case test_ that you wrote for this assignment. |  An edge case was where I checked if the above referenced method could find the first driver and the last driver input in the csv file. |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | This helped a lot in thinking about how I would approach the entire organization of this project as well as breaking down how I would approach creating one method line by line.  |

### CODE REVIEW

File: specs/driver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    driver.driver_id.must_equal driver_id
+
+    driver.must_respond_to :name
+    driver.name.must_equal name
+
+    driver.must_respond_to :vin
+    driver.vin.must_equal vin
+  end
+
+  it "raises an ArgumentError if vin is less than 17 characters" do
```
Comment (droberts-sea) [inline specs/driver_spec.rb:23]: Watch indentation here.

File: specs/driver_spec.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  proc {
+    RideShare::Driver.new(133720003, "allison", "379")
+  }.must_raise ArgumentError
+  end
+
+  it "raises an ArgumentError if vin is more than 17 characters" do
```
Comment (droberts-sea) [inline specs/driver_spec.rb:29]: You've got enough test cases here that it starts to get hard to tell what's doing what. It might be worthwhile to add nested `describe` blocks here to break things up - perhaps one per method.

File: specs/rider_spec.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  end
+
+  it "raises an argument error if the driver doesn't exist" do
+    proc { RideShare::Rider.find_rider(90493928) }.must_raise ArgumentError
+  end
+
+  #for rider_trip_instances
+  it "returns an array of all trip instances for a paticular rider" do
+    new_rider = RideShare::Rider.new(1, "Nina Hintz Sr.", "560.815.3059")
+    new_rider.rider_trip_instances.must_be_instance_of Array
```
Comment (droberts-sea) [inline specs/rider_spec.rb:52]: Here, you should also be testing that the elements in that array are actually `Trip`s, and that they reference the correct `rider_id`.

Moreover, I'd like to see 3 particular test cases here: when there are no trips associated with the rider, when there is exactly one trip, and when there are many trips. Each of those would be a separate `it` block.

File: lib/trip.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      end
+      return all_trips_array
+      # return all the trips in the trip csv
+    end
+
+    #all trip instances with a particular driver id
+    def self.driver_trip_instances(driver_id)
+      driver_trips = []
+      trips = RideShare::Trip.all_trip_info
+      trips.each do |object|
```
Comment (droberts-sea) [inline lib/trip.rb:46]: Could you clean this up a little using an enumerable? Maybe `select` would be a good option.

File: specs/rider_spec.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    new_rider.name.must_equal name
+
+    new_rider.must_respond_to :phone_num
+    new_rider.phone_num.must_equal phone_num
+  end
+
+  #for the self.all_rider_info method
+  it "returns an array or all rider info" do
+    test_class = RideShare::Rider.all_rider_info
+    test_class.must_be_instance_of Array
```
Comment (droberts-sea) [inline specs/rider_spec.rb:26]: I would like to see some coverage here about what's in the array. For example:
- Do you get the number of elements you expect?
- Are all elements instances of `Trip`?
- Did you get the first and last trips out of the CSV?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   Yes! Good work.	|
|   Answer comprehension questions	|   Yes	|
|  **Driver** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   yes |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline comments |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips method to retrieve the associated drivers | yes |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |  mostly - see inline comments  |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   yes |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   yes |
|   Created a method to retrieve all trips by driver id |    yes |
|   Created a method to retrieve all trips by rider id |     yes |
|   **Overall**	|   	|

If I have a complaint, it's about how you name variables and particularly methods. Your method names are quite verbose, to the point that it begins to make your code hard to read. For example, in `Rider` the methods `rider_trip_instances`, `previous_driver_trips` and `all_rider_info` could be `trips`, `drivers` and `all`, respectively. This makes code that calls those methods flow a little better. Moreover, I would argue that the extra verbiage is actually more confusing. If I see a method named `Rider.all`, my assumption is that it's going to load a bunch of data from somewhere and turn it into an Array of `Rider` instances - what else would it do? But if I see a long method name like `Rider.all_rider_info`, I immediately start to worry. Why did the author give it such a long name? Is it doing a bunch of other stuff I didn't think about?

It's a fine line to walk between giving too much and too little with method names, and it's also pretty nitpicky on my part, but it's also something that stood out to me in your code.

Otherwise, your logic looks good, test coverage is on the right track, and your git habits are solid. It's clear you really worked hard on this, and it seems like it's paid off!
